bls-signatures/src/signature.rs

```diff
@@ -139,13 +155,55 @@ impl SignatureProjective {
             )
     }
 
+    /// Verify a list of signatures against a message and a list of public keys
+    /// individually in parallel.
+    ///
+    /// This function first attempts to verify all signatures at once via aggregation.
+    /// If the aggregate verification succeeds, it returns a vector of `true`s.
+    /// If it fails, it falls back to verifying each signature individually in parallel
+    /// to identify the valid and invalid ones.
+    #[cfg(feature = "parallel")]
+    pub fn par_verify_batch<P: AsPubkeyProjective + Sync, S: AsSignatureProjective + Sync>(
+        public_keys: &[&P],
+        signatures: &[&S],
+        message: &[u8],
+    ) -> Result<Vec<bool>, BlsError> {
+        if public_keys.len() != signatures.len() {
+            return Err(BlsError::InputLengthMismatch);
+        }
+
+        if public_keys.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        // First, try to verify by aggregating all public keys and signatures.
+        if SignatureProjective::par_aggregate_verify(public_keys, signatures, message)? {
+            Ok(alloc::vec![true; public_keys.len()])
+        } else {
+            // Use Rayon's parallel iterator to verify each signature concurrently.
+            public_keys
+                .par_iter()
+                .zip(signatures.par_iter())
+                .map(|(pubkey, signature)| {
+                    let pubkey_projective = pubkey.try_as_projective()?;
```

Comment:
We only have 2000 pubkeys per Epoch, if try_as_projective() is not a cheap operation, we can totally pass in PubkeyProjective.

Reply:
Yeah I think it is actually possible to just verify on affine types directly without converting the public keys or signatures to projective. I'll address this in a follow-up.